A Scikit-learn pipeline in Wallaroo

nitbix opened the pull request.

Commit:
add first draft of ML example

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,148 @@
+
+
+
+## Traning vs Inference
```
`Training` not `Traning`

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,148 @@
++++
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a sample pipeline to classify images from the [MNIST dataset](ihttp://yann.lecun.com/exdb/mnist/), using PCA and a regression tree in scikit-learn. As a prerequisite, you should be familiar with some notions of machine learning ...
+Everything has been implemented in the [current version of Wallaroo (0.2.2)](https://github.com/WallarooLabs/wallaroo/tree/0.2.2). The full code can be found on [GitHub](https://github.com/WallarooLabs/wallaroo_blog_examples/tree/master/sklearn-example). If you have any technical questions that this post didn't answer, or if you have any suggestions, please get in touch at [hello@WallarooLabs.com ...
```
0.2.2 is not the current version.

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,148 @@
+
+## Running our application
+
+To run our application, we need to follow these steps:
```
I think "running the application" makes more sense in the README for the example. Given you have to go there and clone it etc. It doesnt make much sense here.

nitbix replied:
We have this in other blog entries, so for consistency perhaps the best thing is to have it in both places?

SeanTAllen commented on content/post/mnist-example.md:
````diff
@@ -0,0 +1,148 @@
+    return ab.build()
+```
+
+Note how we are loading our model in `application_setup` and making them global.  This is because loading a pickled scikit-learn model could be a potentially expensive operation, and we want to ensure we do it only once per worker. By doing so in the application_setup, the models are only loaded during initialization, and made available to the computations via the `global` keyword.
````
This is good but I think you should explain what the flow of data that the application builder is setting up,

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,148 @@
+
+### Encoder
+
+The encoder simply packs the result into a string and sends it across the wire.
```
it doesn't though. its using a framed protocol.

either the comment or the code should be updated accordingly.

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,148 @@
+author = "amosca"
++++
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a sample pipeline to classify images from the [MNIST dataset](ihttp://yann.lecun.com/exdb/mnist/), using PCA and a regression tree in scikit-learn. As a prerequisite, you should be familiar with some notions of machine learning ...
```
I think you need to explain what the MNIST dataset is.

I think you should explain the basics of some of these notions.

Commit:
First round of comments on sklearn-example

- Add some background information
- Fix some errors

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
+    "tutorial"
+]
+categories = [
+    "Machine Learning"
```
i think this should be `Examples`

nitbix replied:
both maybe?

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
++++
+title = "A Scikit-learn pipeline in Wallaroo"
+draft = false
+date = 2017-11-16T00:00:00Z
```
Date needs to be corrected. I mention it more because it took me a while to find the post on the netlify preview then anything else (I was looking for the preview at the top of the site with the newest posts).

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
+author = "amosca"
++++
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a sample pipeline to classify images from the [MNIST dataset](ihttp://yann.lecun.com/exdb/mnist/), using PCA and a Logistic Regression in scikit-learn. As a prerequisite, you should be familiar with some notions of machine lear ...
```
Why are `many applications with streaming data are either applying machine learning or have a very good use case for it`? This needs more context for the uninformed reader.

It would be great if you could expand on "some notions of machine learning". As someone who knows a very tiny bit about ML, I have no clue if I know enough or not for this example (unless "some notions" means "PCA and logistic regression" from the sentence prior, in which case I didn't make that connection).

Additionally, the "notions about Wallaroo pipelines" should link to either an earlier blog post or the documentation at a minimum. It would be great if we can add in a line or two summary of what Wallaroo is/allows. Maybe something like "wallaroo is our streaming platform that allows developers to write in either Python or Go and autoscale easily."

dipinhora replied:
It would also be great to have a one or two line summary of `scikit-learn` along with a link to it for folks who want to read more.

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
+
+## The model
+
+One of the simplest models for classifying digits is a [Logistic Regression](https://en.wikipedia.org/wiki/Logistic_regression) on the numeric values of each pixel. An improvement to this simple model is to add a [PCA](https://en.wikipedia.org/wiki/Principal_component_analysis) preprocessing that presents the transformation with the most information content as an input to the classifier.
```
It would be great if we can write out `Principal Component Analysis (PCA) so people who aren't as familiar with ML know what PCA stands for.

I don't know what the goal for this post is (or who the audience is), but it would be great if you are able to include a 1 or 2 line explanation of how LR and/or PCA do what they do and why classification is important.

Also, this section is talking about `classifying digits` but the previous section said `Each of these images has been classified`. Why would we need to classify digits if the images are already classified?

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
+
+## Running our application
+
+To run our application, we need to follow these steps:
```
Can we elaborate on what folks will see when they run the application so they know what to expect?

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,155 @@
+
+## Next steps
+
+There are obvious limitations to this basic example. For instance, there is no partitioning. And we of course realize that MNIST isn't a useful dataset beyond examples. A lot of extra functionality can be added to production-level code, but for the purpose of illustrating how to run scikit-learn algorithms in Wallaroo, we preferred to narrow the focus and reduce distractions.
```
I'm not sure the limitations are "obvious" nor that people will "of course realize" that MNIST isn't useful beyond examples. That would depend on peoples knowledge of Wallaroo, ML, and/or the MNIST dataset. It would be great if we can explain why MNIST isn't useful beyond examples (one or two sentences would be great). Similarly, we can point out that there are additional wallaroo features not used (such as partitioning, etc) and link to them for further reading by folks.

Commit:
Add more meat to the introduction in mnist-example

aturley commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,158 @@
+
+Whilst it would seem that machine learning is taking over the world, a lot of the attention has been focused towards researching new methods and applications, and how to make a single model faster. At Wallaroo Labs we believe that, in order to make the benefits of machine learning truly ubiquitous, there needs to be a significant improvement in how we put those impressive models into production.  ...
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a machine learning inside Wallaroo, our high performance stream processing engine, to classify images from the [MNIST dataset](ihttp://yann.lecun.com/exdb/mnist/), using a basic two-stage model in Python. Whilst recognizing han ...
```
The link to the MINST dataset (`ihttp://yann.lecun.com/exdb/mnist/`) seems to be incorrect. I'm guessing it should be `http://yann.lecun.com/exdb/mnist/`

aturley commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,158 @@
+        pca = pickle.load(f)
+    with open('logistic.model', 'r') as f:
+        logistic = pickle.load(f)
+    global pca
```
These `global` statements should come before the code assigns to the variables. The code will work, but if you do something like this in a normal Python program you'll actually get a warning about it.

aturley commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,158 @@
+
+## Creating the models
+
+Even though we have said that the focus will be on inference, we still need to create some models to be able to use them. In this case, all the code for training is contained in `train_models.py`. We invite you to take a detailed look at it, but for the sake of this blog entry we only need to know that it is training a PCA for data preprocessing and a logistic regression for classification. The t ...
```
It would be good to have a link to `train_models.py`.

aturley replied:
Also, "to two separate file" should be "to two separate files".

Commit:
Add more background

- What is Wallaroo section
- Check it out section
- Fix some issues found in the PR

Commit:
remove Go survey

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,174 @@
+
+Whilst it would seem that machine learning is taking over the world, a lot of the attention has been focused towards researching new methods and applications, and how to make a single model faster. At Wallaroo Labs we believe that, in order to make the benefits of machine learning truly ubiquitous, there needs to be a significant improvement in how we put those impressive models into production.  ...
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a machine learning inside Wallaroo, our high performance stream processing engine, to classify images from the [MNIST dataset](http://yann.lecun.com/exdb/mnist/), using a basic two-stage model in Python. Whilst recognizing hand ...
```
It seems there's a word missing in

> In this example, we will explore how we can build a machine learning inside Wallaroo

maybe it was supposed to be

> In this example, we will explore how we can build a machine learning **_application_** inside Wallaroo

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,174 @@
+
+Nowadays, many applications with streaming data are either applying machine learning or have a very good use case for it. In this example, we will explore how we can build a machine learning inside Wallaroo, our high performance stream processing engine, to classify images from the [MNIST dataset](http://yann.lecun.com/exdb/mnist/), using a basic two-stage model in Python. Whilst recognizing hand ...
+
+Everything has been implemented in the [current version of Wallaroo (1.4.0)](https://github.com/WallarooLabs/wallaroo/tree/0.4.0). The full code can be found on [GitHub](https://github.com/WallarooLabs/wallaroo_blog_examples/tree/master/sklearn-example). If you have any technical questions that this post didn't answer, or if you have any suggestions, please get in touch at [hello@WallarooLabs.com ...
```
Typo: `1.4.0` should be `0.4.0`.

dipinhora commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,174 @@
+
+Everything has been implemented in the [current version of Wallaroo (1.4.0)](https://github.com/WallarooLabs/wallaroo/tree/0.4.0). The full code can be found on [GitHub](https://github.com/WallarooLabs/wallaroo_blog_examples/tree/master/sklearn-example). If you have any technical questions that this post didn't answer, or if you have any suggestions, please get in touch at [hello@WallarooLabs.com ...
+
+We've been working on our processing engine, [Wallaroo](https://github.com/wallaroolabs/wallaroo/tree/release) for just under two years now. Our goal has been to make it as easy to build fast, scale-independent applications for processing data. When we open sourced Wallaroo last year we provided an API that let developers create applications using [Python](https://blog.wallaroolabs.com/2017/10/go ...
```
This paragraph doesn't seem to flow from the previous one. Not sure how to change things to make it flow better but wanted to mention it.

It would be better to point people to the blog post regarding the new Python API that doesn't use objects as that's what release `0.4.0` has.

Additionally, it might make sense to mention that Wallaroo also provides a `Go` API.

dipinhora commented on content/post/mnist-example.md:
````diff
@@ -0,0 +1,174 @@
+    return ab.build()
+```
+
+Note how we are loading our model in `application_setup` and making them global.  This is because loading a pickled scikit-learn model could be a potentially expensive operation, and we want to ensure we do it only once per worker. By doing so in the application_setup, the models are only loaded during initialization, and made available to the computations via the `global` keyword.
````
The blog sometimes uses `scikit-learn` and sometimes uses `sklearn`. It would be better if one variant is used consistently.

nitbix replied:
the convention is that sklearn is only used in imports and links. scikit-learn is always used in "conversation"

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,174 @@
+    "tutorial"
+]
+categories = [
+    "Machine Learning",
```
remove this

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,174 @@
+
+There are obvious limitations to this basic example. For instance, there is no partitioning. And we of course realize that MNIST isn't a useful dataset beyond examples. A lot of extra functionality can be added to production-level code, but for the purpose of illustrating how to run scikit-learn algorithms in Wallaroo, we preferred to narrow the focus and reduce distractions.
+
+If you’d like to see the full code, its available on [GitHub](https://github.com/WallarooLabs/wallaroo_blog_examples/tree/master/sklearn-example). If you would like to ask us more in-depth technical questions, or if you have any suggestions, please get in touch via [our mailing list](https://groups.io/g/wallaroo) or [our IRC channel](https://webchat.freenode.net/?channels=#wallaroo).
```
this paragraph is the same content as first in check out out. i think you can kill this one.

Commit:
Address final comments

Commit:
fix date

SeanTAllen commented on content/post/mnist-example.md:
```diff
@@ -0,0 +1,175 @@
+
+There are obvious limitations to this basic example. For instance, there is no partitioning. And we of course realize that MNIST isn't a useful dataset beyond examples. A lot of extra functionality can be added to production-level code, but for the purpose of illustrating how to run scikit-learn algorithms in Wallaroo, we preferred to narrow the focus and reduce distractions.
+
+## Check It Out
```
this can be deleted. and the following content becomes part of next steps

Commit:
remove "check it out"

SeanTAllen committed:
Spelling and grammar fixes
